[Bug]: Upgrade fails with `npx storybook@latest upgrade`

### Describe the bug

Trying to upgrade from `v8.4.7` to `v8.5.0` and get this error. Something is wrong with dependecnies. There is already closed similar issue (#29169).

```bash
$ npx storybook@latest upgrade
WARN An issue occurred while trying to find dependencies metadata using npm.
╭───────────────────────────────────────────────────────────────╮
│                                                               │
│   Upgrading Storybook from version 0.0.0 to version 8.5.0..   │
│                                                               │
╰───────────────────────────────────────────────────────────────╯
info Checking for upgrade blockers...
No blockers found.

info Updating dependencies in package.json..
Installing dependencies...

npm ERR! code ERESOLVE
npm ERR! ERESOLVE could not resolve
npm ERR!
npm ERR! While resolving: my-package@1.0.11
npm ERR! Found: @storybook/react@8.4.7
npm ERR! node_modules/@storybook/react
npm ERR!   dev @storybook/react@"^8.5.0" from the root project
npm ERR!   @storybook/react@"8.4.7" from @storybook/react-vite@8.4.7
npm ERR!   node_modules/@storybook/react-vite
npm ERR!     dev @storybook/react-vite@"^8.5.0" from the root project
npm ERR!
npm ERR! Could not resolve dependency:
npm ERR! dev @storybook/react@"^8.5.0" from the root project
npm ERR!
npm ERR! Conflicting peer dependency: @storybook/test@8.5.0
npm ERR! node_modules/@storybook/test
npm ERR!   peerOptional @storybook/test@"8.5.0" from @storybook/react@8.5.0
npm ERR!   node_modules/@storybook/react
npm ERR!     dev @storybook/react@"^8.5.0" from the root project
npm ERR!
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
```

### Reproduction link

https://stackblitz.com/edit/vitejs-vite-izegxbtn

### Reproduction steps

_No response_

### System

```bash
$ npx storybook@latest info

Storybook Environment Info:

  System:
    OS: Windows 11 10.0.26100
    CPU: (24) x64 AMD Ryzen 9 3900XT 12-Core Processor
  Binaries:
    Node: 22.10.0 - C:\Program Files\nodejs\node.EXE
    npm: 9.7.2 - C:\Program Files\nodejs\npm.CMD <----- active
  Browsers:
    Edge: Chromium (131.0.2903.86)
  npmPackages:
    @storybook/addon-essentials: ^8.4.7 => 8.4.7
    @storybook/addon-interactions: ^8.4.7 => 8.4.7
    @storybook/addon-onboarding: ^8.4.7 => 8.4.7
    @storybook/blocks: ^8.4.7 => 8.4.7
    @storybook/react: ^8.4.7 => 8.4.7
    @storybook/react-vite: ^8.4.7 => 8.4.7
    @storybook/test: ^8.4.7 => 8.4.7
    eslint-plugin-storybook: ^0.11.0 => 0.11.0
    storybook: ^8.4.7 => 8.4.7
```

### Additional context

_No response_

Same issue, 8.4.7 to 8.5.1

`npm error code ERESOLVE
npm error ERESOLVE could not resolve
npm error
npm error While resolving: my-package@6.2.0
npm error Found: @storybook/react@8.4.7
npm error node_modules/@storybook/react
npm error   dev @storybook/react@"^8.5.1" from the root project
npm error   @storybook/react@"8.4.7" from @storybook/preset-react-webpack@8.4.7
npm error   node_modules/@storybook/preset-react-webpack
npm error     @storybook/preset-react-webpack@"8.4.7" from @storybook/react-webpack5@8.4.7
npm error     node_modules/@storybook/react-webpack5
npm error       dev @storybook/react-webpack5@"^8.5.1" from the root project
npm error   1 more (@storybook/react-webpack5)
npm error
npm error Could not resolve dependency:
npm error dev @storybook/react@"^8.5.1" from the root project
npm error
npm error Conflicting peer dependency: @storybook/test@8.5.1
npm error node_modules/@storybook/test
npm error   peerOptional @storybook/test@"8.5.1" from @storybook/react@8.5.1
npm error   node_modules/@storybook/react
npm error     dev @storybook/react@"^8.5.1" from the root project
npm error
npm error Fix the upstream dependency conflict, or retry
npm error this command with --force or --legacy-peer-deps
npm error to accept an incorrect (and potentially broken) dependency resolution.
Error: Error: Command failed with exit code 1: npm install
`

There's some kind of dependency race condition and in my case I just run `npm i --force` after upgrading `storybook`. Nevertheless this condition breaks `npx storybook upgrade` as it exits with error not being able to install upgraded storybook.

For reference error that I got when upgrading 8.5.1 to 8.5.2 says:
```
npm error ERESOLVE could not resolve
npm error
npm error While resolving: ....
npm error Found: @storybook/addon-actions@8.5.1
npm error node_modules/@storybook/addon-actions
npm error   dev @storybook/addon-actions@"^8.5.2" from the root project
npm error   @storybook/addon-actions@"8.5.1" from @storybook/addon-essentials@8.5.1
npm error   node_modules/@storybook/addon-essentials
npm error     dev @storybook/addon-essentials@"^8.5.2" from the root project
npm error
npm error Could not resolve dependency:
npm error dev @storybook/addon-actions@"^8.5.2" from the root project
npm error
npm error Conflicting peer dependency: storybook@8.5.2
npm error node_modules/storybook
npm error   peer storybook@"^8.5.2" from @storybook/addon-actions@8.5.2
npm error   node_modules/@storybook/addon-actions
npm error     dev @storybook/addon-actions@"^8.5.2" from the root project
```

--- a/code/lib/cli-storybook/src/upgrade.ts
+++ b/code/lib/cli-storybook/src/upgrade.ts
@@ -250,6 +250,27 @@ export const doUpgrade = async ({
     const upgradedDependencies = toUpgradedDependencies(packageJson.dependencies);
     const upgradedDevDependencies = toUpgradedDependencies(packageJson.devDependencies);
 
+    // Users struggle to upgrade Storybook with npm because of conflicting peer-dependencies
+    // GitHub Issue: https://github.com/storybookjs/storybook/issues/30306
+    // Solution: Remove all Storybook packages (except 'storybook') from the package.json and install them again
+    if (packageManager.type === 'npm') {
+      await packageManager.removeDependencies(
+        {
+          skipInstall: false,
+        },
+        [...upgradedDependencies, ...upgradedDevDependencies]
+          .map((dep) => {
+            const lastAtIndex = dep.lastIndexOf('@');
+            return lastAtIndex > 0 ? dep.slice(0, lastAtIndex) : dep;
+          })
+          // We don't want to remove the storybook package from the package.json
+          // because third-party packages which we don't remove may depend on it via peer-dependency requirements
+          .filter((dep) => dep !== 'storybook')
+      );
+
+      await packageManager.installDependencies();
+    }
+
     logger.info(`Updating dependencies in ${picocolors.cyan('package.json')}..`);
     if (upgradedDependencies.length > 0) {
       await packageManager.addDependencies(

